File: eco-core/core-plugin/src/main/java/com/willfp/ecoarmor/sets/ArmorSets.java
package com.willfp.ecoarmor.sets;


import com.google.common.collect.BiMap;
import com.google.common.collect.HashBiMap;
import com.google.common.collect.ImmutableList;
import com.willfp.eco.core.config.ConfigUpdater;
import com.willfp.eco.core.config.JSONConfig;
import com.willfp.ecoarmor.EcoArmorPlugin;
import lombok.experimental.UtilityClass;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

@UtilityClass
public class ArmorSets {
    /**
     * Registered armor sets.
     */
    private static final BiMap<String, ArmorSet> BY_NAME = HashBiMap.create();

    /**
     * Get all registered {@link ArmorSet}s.
     *
     * @return A list of all {@link ArmorSet}s.
     */
    public static List<ArmorSet> values() {
        return ImmutableList.copyOf(BY_NAME.values());
    }

    /**
     * Get {@link ArmorSet} matching name.
     *
     * @param name The name to search for.
     * @return The matching {@link ArmorSet}, or null if not found.
     */
    @Nullable
    public static ArmorSet getByName(@NotNull final String name) {
        return BY_NAME.get(name);
    }

    /**
     * Update all {@link ArmorSet}s.
     *
     * @param plugin Instance of EcoArmor.
     */
    @ConfigUpdater
    public static void update(@NotNull final EcoArmorPlugin plugin) {
        for (ArmorSet set : values()) {
            removeSet(set);
        }

        for (JSONConfig setConfig : plugin.getEcoArmorJson().getSubsections("sets")) {
            addNewSet(new ArmorSet(setConfig, plugin));
        }
    }

    /**
     * Add new {@link ArmorSet} to EcoArmor.
     *
     * @param set The {@link ArmorSet} to add.
     */
    public static void addNewSet(@NotNull final ArmorSet set) {
        BY_NAME.remove(set.getName());
        BY_NAME.put(set.getName(), set);
    }

    /**
     * Remove {@link ArmorSet} from EcoArmor.
     *
     * @param set The {@link ArmorSet} to remove.
     */
    public static void removeSet(@NotNull final ArmorSet set) {
        BY_NAME.remove(set.getName());
    }
}

File: eco-core/core-plugin/src/main/kotlin/com/willfp/ecoarmor/sets/ArmorSets.kt
package com.willfp.ecoarmor.sets

import com.google.common.collect.ImmutableList
import com.willfp.eco.core.config.interfaces.Config
import com.willfp.eco.core.registry.Registry
import com.willfp.libreforge.loader.LibreforgePlugin
import com.willfp.libreforge.loader.configs.ConfigCategory
import com.willfp.libreforge.loader.configs.LegacyLocation

object ArmorSets : ConfigCategory("set", "sets") {
    /**
     * Registered armor sets.
     */
    private val registry = Registry<ArmorSet>()

    private var cachedValues: List<ArmorSet> = emptyList()

    override val legacyLocation = LegacyLocation(
        "ecoarmor.yml",
        "sets"
    )

    /**
     * Get all registered [ArmorSet]s.
     *
     * @return A list of all [ArmorSet]s.
     */
    @JvmStatic
    fun values(): List<ArmorSet> {
        return cachedValues
    }

    /**
     * Get [ArmorSet] matching ID.
     *
     * @param name The name to search for.
     * @return The matching [ArmorSet], or null if not found.
     */
    @JvmStatic
    fun getByID(name: String): ArmorSet? {
        return registry[name]
    }

    override fun clear(plugin: LibreforgePlugin) {
        registry.clear()
        cachedValues = emptyList()
    }

    override fun acceptConfig(plugin: LibreforgePlugin, id: String, config: Config) {
        registry.register(ArmorSet(id, config))
    }

    override fun afterReload(plugin: LibreforgePlugin) {
        cachedValues = ImmutableList.copyOf(registry.values())
    }
}
